src/utils/gasPriceManager.ts

```diff
@@ -0,0 +1,450 @@
+        if (maxFeePerGas === undefined) {
+            this.logger.warn("maxFeePerGas is undefined, using fallback value")
+            try {
+                maxFeePerGas =
+                    (await this.getNextBaseFee(this.publicClient)) +
+                    maxPriorityFeePerGas
+            } catch (e) {
+                this.logger.error("failed to get fallback maxFeePerGas")
+                sentry.captureException(e)
+                throw e
+            }
+        }
+
+        if (maxPriorityFeePerGas === 0n) {
+            maxPriorityFeePerGas = maxFeePerGas / 200n
+        }
+
+        return { maxFeePerGas, maxPriorityFeePerGas }
+    }
+
+    private saveMaxFeePerGas(gasPrice: bigint, timestamp: number) {
+        const queue = this.queueMaxFeePerGas
+        const last = queue.length > 0 ? queue[queue.length - 1] : null
+
+        if (!last || timestamp - last.timestamp >= 1000) {
+            if (queue.length >= this.maxQueueSize) {
+                queue.shift()
+            }
+            queue.push({ maxFeePerGas: gasPrice, timestamp })
+        } else if (gasPrice < last.maxFeePerGas) {
+            last.maxFeePerGas = gasPrice
+            last.timestamp = timestamp
+        }
+    }
+
+    private saveMaxPriorityFeePerGas(gasPrice: bigint, timestamp: number) {
+        const queue = this.queueMaxPriorityFeePerGas
+        const last = queue.length > 0 ? queue[queue.length - 1] : null
+
+        if (!last || timestamp - last.timestamp >= 1000) {
+            if (queue.length >= this.maxQueueSize) {
+                queue.shift()
+            }
+            queue.push({ maxPriorityFeePerGas: gasPrice, timestamp })
+        } else if (gasPrice < last.maxPriorityFeePerGas) {
+            last.maxPriorityFeePerGas = gasPrice
+            last.timestamp = timestamp
+        }
+    }
+
+    private saveGasPrice(gasPrice: GasPriceParameters, timestamp: number) {
+        return new Promise<void>((resolve) => {
+            this.saveMaxFeePerGas(gasPrice.maxFeePerGas, timestamp)
+            this.saveMaxPriorityFeePerGas(
+                gasPrice.maxPriorityFeePerGas,
+                timestamp
+            )
+            resolve()
+        })
+    }
+
+    public async getGasPrice(): Promise<GasPriceParameters> {
+        let maxFeeFloor = 0n
+        let maxPriorityFeeFloor = 0n
+
+        if (this.chain.id === chains.dfk.id) {
+            maxFeeFloor = 5_000_000_000n
+            maxPriorityFeeFloor = 5_000_000_000n
+        }
+
+        if (
+            this.chain.id === chains.polygon.id ||
+            this.chain.id === chains.polygonMumbai.id
+        ) {
+            const polygonEstimate = await this.getPolygonGasPriceParameters()
+            if (polygonEstimate) {
+                const gasPrice = this.bumpTheGasPrice({
+                    maxFeePerGas: polygonEstimate.maxFeePerGas,
+                    maxPriorityFeePerGas: polygonEstimate.maxPriorityFeePerGas
+                })
+                this.saveGasPrice(
```

Comment:
why are we doing saveGasPrice 3 times in this function instead of abstracting the main implementation into an `innerGetGasPrice` and calling this right after

Reply:
it makes things explicit calculating gas price is separate from saving it

Reply:
right, i mean you could have `getGasPrice` which underneath just calls two functions, `innerGetGasPrice` and `saveGasPrice`